src/details/ArborX_DetailsTreeTraversal.hpp

```diff
@@ -88,7 +93,12 @@ struct TreeTraversal<BVH, Predicates, Callback, SpatialPredicateTag>
   KOKKOS_FUNCTION void operator()(int queryIndex) const
   {
     auto const &predicate = Access::get(_predicates, queryIndex);
+    search(predicate);
+  }
 
+  template <typename Query>
+  KOKKOS_FUNCTION void search(Query const &predicate) const
+  {
```

Comment:
Instead of introducing a function, we could only have a thread variant, and grab the query inside the parallel_loop.